### ISSUE TITLE
fgMarkAddressExposedLocals issue with wider indirs

### ISSUE THREAD

**Issue by mikedn** [dotnet/coreclr#16472]
If the address of a struct field is taken and used by an indir node that is wider than the field then `fgMarkAddressExposedLocals` marks only the field as being address taken, it should probably mark the entire struct variable.

```c#
struct foobar {
    public byte b1, b2, b3, b4;
    public foobar(int x) {
        b1 = 1;
        b2 = 2;
        b3 = 3;
        b4 = 4;
    }
}
[MethodImpl(MethodImplOptions.NoInlining)]
static int Test() {
    foobar f = new foobar(42);
    return Unsafe.As<byte, int>(ref f.b1);
}
```
generates
```asm
G_M55888_IG01:
       50                   push     rax
G_M55888_IG02:
       C644240401           mov      byte  ptr [rsp+04H], 1
       8B442404             mov      eax, dword ptr [rsp+04H] ; reads 3 bytes of garbage data
G_M55888_IG03:
       4883C408             add      rsp, 8
       C3                   ret
```

Extracted from https://github.com/dotnet/coreclr/issues/16210#issuecomment-367093493.

It's worth noting that this does not happen if pointers are used instead of `Unsafe.As`. That's due to another `fgMarkAddressExposedLocals` issue: the pointer based code contains a cast (from byref to native int) that `fgMarkAddressExposedLocals` does not understand. I was actually looking at getting the pointer version to work when I ran into this issue.

category:correctness
theme:morph
skill-level:expert
cost:medium

**Comment by mikedn** [comment]
As mentioned in https://github.com/dotnet/coreclr/issues/16210#issuecomment-367110476 it may better to make `fgMarkAddressExposedLocals` track the indir width and offset instead of (or in addition to) the current vague `AXC_IndWide`/`AXC_AddrWide` values.

The main problem in getting this to work right (and less complex than it currently is) seems to be assignments and blk/obj nodes. It is possible to construct IR like
```
[000026] ------------              *  STMT      void  (IL 0x018...0x024)
[000024] -------N----              |  /--*  IND       struct
[000021] ---------U--              |  |  \--*  CAST      long <- ulong <- ulong
[000020] L-----------              |  |     \--*  ADDR      long  
[000019] ------------              |  |        \--*  LCL_VAR   struct V01 loc1         
[000025] -A------R---              \--*  ASG       struct (copy)
[000023] -------N----                 \--*  BLK(80)   struct
[000018] ---------U--                    \--*  CAST      long <- ulong <- ulong
[000017] L-----------                       \--*  ADDR      long  
[000016] ------------                          \--*  LCL_VAR   struct V00 loc0         
```
where the indir node doesn't have an explicit width and you need to look across the assignment to figure it out. It's not clear to me why the JIT generates an IND node and not an OBJ node of the appropriate size.

**Comment by RussKeldorph** [comment]
@dotnet/jit-contrib

**Comment by mikedn** [comment]
@RussKeldorph @AndyAyersMS Do really want to put this in 2.1? I'm not sure that lookahead will work here because there may be an arbitrary chain of fields between the indir and the lclvar. And the proper fix - tracking the indir width and offset accurately - seems to be quite a bit of work to get right. FWIW I prototyped something for fun but it will take a while to finish it.

**Comment by RussKeldorph** [comment]
@mikedn Thanks for the bug and the additional info.  We generally like to fix silent bad codegen ASAP so I started with 2.1.0.  Things like unsafe, not a regression, and difficulty to fix certainly lower its chances, so we may punt on this.

**Comment by AndyAyersMS** [comment]
I think moving this out of 2.1 is reasonable.

We addressed the one externally reported case of this #16210 via #16305, but the mechanism used there is becoming increasingly clunky.

I think we need time to step back and reconsider how we handle this analysis in general. I expect it will be challenging to find a fix that is correct and not overly conservative. 

There are other ways to get at the underlying issue and reasons to believe they'll become more common in the future, but for right now this is a fairly rare case.

**Comment by mikedn** [comment]
> I think we need time to step back and reconsider how we handle this analysis in general

IMO pushing an "address context" down the tree, while it seems to make sense at the first look, is rather unintuitive and has at least 2 problems that seem unavoidable:
* you can't push down a different address context for each sub-tree of an n-ary operator (e.g. CALL), at least not in the current tree walker based implementation
* you can't handle the case of getting the offset of a field by subtracting pointers (#16527)

Pushing up the tree a simple "points to" value consisting of a lclNum and a byte offset seems to be simpler and avoids the above problems. Basically:
 
```
LCLVAR(x) -> val1 = (isAddress: false, lclNum: x, offset: 0)
ADDR(val1) -> val2 = (isAddress: true, lclNum: x, offset: 0)
FIELD(val2, ofs) -> val3 = (isAddress: false, lclNum: x, offset: ofs)
IND(val2) -> val4 = (nothing) and if the indir is wider than x then mark x address taken
UNKNOWN(val2) -> val5 = (nothing) and mark x address taken

and so on, the usual evaluation rules you'd expect for something like this
```

Prototype here: https://github.com/mikedn/coreclr/commit/ad9ea1274be375e63fb90852c88092363cd0e7a9#diff-bd2d23afdb1936d1982f699f760429d2
Took a bit of massaging but I managed to bring diffs down to 0. From here other fixes and improvements can be done.

**Comment by mikedn** [comment]
@AndyAyersMS Does the above prototype looks like a reasonable start?

**Comment by mikedn** [comment]
@RussKeldorph You can assign to me, fix in progress in #19621

**Comment by mikedn** [comment]
Fixed in #19621
